### PR TITLE
Test Anything Protocol (TAP)

### PR DESCRIPTION
**Pull request by sebastianbergmann**
> [TAP, the Test Anything Protocol](https://testanything.org/tap-version-14-specification.html), is a simple text-based interface between testing modules in a test harness. It decouples the reporting of errors from the presentation of the reports.

In other words: TAP is a plain text log file format for test results that is agnostic to the test framework or test runner used.

Earlier versions of PHPUnit had support for TAP, but this support was removed some time ago.

[I think that "output in CI pipelines" is a separate use case and as such should be addressed using a new/separate output printer](https://github.com/sebastianbergmann/phpunit/issues/5883#issuecomment-2219890403) and I think that the TAP format might be suitable for that purpose.

So far, this PR implements

- an XML configuration option for logging test results in TAP format to a file
- a `--tap` CLI option for replacing the default test runner output with output in TAP format
- a `--log-tap` CLI option for logging test results in TAP format to a file
- a minimal implementation of a TAP logger (which does not handle all possible outcomes and issues a test may have)

Here is an example of what the TAP output currently looks like:

https://github.com/sebastianbergmann/phpunit/blob/4cf45887d298eb07b2ee44a81aaa1819009a3cb4/tests/end-to-end/logging/tap.phpt#L14-L109

@TimWolla @edorian @Bilge What do you think?

I will not pursue this further if the consensus is that TAP does not solve the "output in CI pipelines" use case.

### CODE REVIEW

File: tests/end-to-end/logging/tap.phpt [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+<?php declare(strict_types=1);
+$_SERVER['argv'][] = '--do-not-cache-result';
+$_SERVER['argv'][] = '--no-configuration';
+$_SERVER['argv'][] = '--tap';
+$_SERVER['argv'][] = __DIR__ . '/../_files/OutcomesAndIssuesTest.php';
+
+require_once __DIR__ . '/../../bootstrap.php';
+
+(new PHPUnit\TextUI\Application)->run($_SERVER['argv']);
+--EXPECTF--
```
Comment (TimWolla) [inline tests/end-to-end/logging/tap.phpt:13]: FWIW: PHPT tests themselves are intended to be valid PHP files, so I would recommend adding the closing PHP tag.

**Review by TimWolla — COMMENTED** [review]
Thank you. I was primarily looking for an indicator of which specific test is currently running and the “successfully prepared …” comments nicely solve this.

One note that may or may not be related to this being a WIP: It appears that the “ok …” output is not working for PHPT tests, the “not ok” works correctly.

**Comment by edorian** [comment]
I've only had second hand problems with this as it was not possible for a colleague to see which test was hanging in CI. They were using `--testdox`, which incidentally solved this problem, and when that went away, the PHPUnit 11 update needed a solution.

A **very** simple PHPUnit event listener [1] was good enough to solve this problem for us. For me, this particular problem is a bit of a homemade edge case anyway.

Over the years I've had a few cases where a "--give-me-each-test-name-before-starting-and-print-its-runtime-afterwards" option would have been handy, but that may be beside the point :)

In this particular case, the `--tap' output would have done the job as well. I also forgot about `--debug`, which would also be fine, but a bit more tedious to read.

So all I can say is that I'm happy with what PHPUnit gives me here and I don't have a strong need for TAP output, but it would work for our current "issue".

[1]

<details>
<summary>Print current test</summary>
<pre><code>
class PrintCurrentTest implements \PHPUnit\Runner\Extension\Extension
{
    public function bootstrap(
        \PHPUnit\TextUI\Configuration\Configuration $configuration,
        \PHPUnit\Runner\Extension\Facade $facade,
        \PHPUnit\Runner\Extension\ParameterCollection $parameters,
    ): void {
        $facade->registerSubscriber(new class implements PHPUnit\Event\Test\PreparedSubscriber {
            public function notify(\PHPUnit\Event\Test\Prepared $event): void
            {
                $name = explode('/', $event->test()->name()); // PHPT test, so the test name is a file path
                fwrite(STDOUT, "\nRunning: " . $name[count($name) - 2]);
            }
        });
    }
}
</code></pre>
</details>

**Comment by Bilge** [comment]
Let's be clear, this isn't just for CI. Having immediate output is useful in any environment, including locally. To that end, I'm not really sure why you're married to the classic PHPUnit output; to me it feels outdated and should be replaced, but I'm not here to convince you of that.

As far as TAP is concerned, of course it solves the immediate output problem, though I've never been a fan of its style.

Counter-point: [Pip](https://github.com/ScriptFUSION/PHPUnit-Immediate-Printer). I just resurrected this printer from the PHPUnit 5/6 days since the new event framework makes it much easier to write and maintain, and of course, because this is the first version without an immediate printer built-in. I don't think its classic format I cooked up seven years ago looks much better than TAP, but I will now begin designing a more modern theme, probably modelled on [Collision](https://github.com/nunomaduro/collision). I would be happy to submit it for inclusion in PHPUnit, either directly (e.g. imported into this repo or as a Composer dependency) or indirectly (e.g. just a link in the docs), if you like.

File: tests/end-to-end/logging/tap.phpt (hunk 1 above)
Comment (sebastianbergmann) [reply tests/end-to-end/logging/tap.phpt:13]: I know that https://qa.php.net/write-test.php has

> The FILE section is used as the body of the .php file, so don't forget to open and close your php tags.

but I do not consider a PHP file without a closing PHP tag to be invalid and have never run into problems related to not using closing PHP tags in PHPT tests (that I am aware of).

**Comment by Bilge** [comment]
@sebastianbergmann Since you closed this, what _is_ your plan to support real-time test output going forward?

**Comment by sebastianbergmann** [comment]
I do not plan to support real-time output, at least not at the moment.
